### PR TITLE
FormAuthenticationMechanism customisation

### LINKED ISSUE #14303: More config options for FormAuthenticationMechanism
**Description**
The `FormAuthenticationMechanism` can be great to use to perform form based auth, however it lacks some personalisation features, the post location for example is hard coded (/j_security_check), the username/password too 

```java
public class FormAuthenticationMechanism implements HttpAuthenticationMechanism {

    private static final Logger log = Logger.getLogger(FormAuthenticationMechanism.class);

    public static final String DEFAULT_POST_LOCATION = "/j_security_check";

    private final String loginPage;
    private final String errorPage;
    private final String postLocation = DEFAULT_POST_LOCATION;
    private final String locationCookie = "quarkus-redirect-location";
    private final String landingPage;
    private final boolean redirectAfterLogin;

    private final PersistentLoginManager loginManager;

    public FormAuthenticationMechanism(String loginPage, String errorPage, String landingPage, boolean redirectAfterLogin,
            PersistentLoginManager loginManager) {
        this.loginPage = loginPage;
        this.errorPage = errorPage;
        this.landingPage = landingPage;
        this.redirectAfterLogin = redirectAfterLogin;
        this.loginManager = loginManager;
    }

    public Uni<SecurityIdentity> runFormAuth(final RoutingContext exchange,
            final IdentityProviderManager securityContext) {
        exchange.request().setExpectMultipart(true);
        return Uni.createFrom().emitter(new Consumer<UniEmitter<? super SecurityIdentity>>() {
            @Override
            public void accept(UniEmitter<? super SecurityIdentity> uniEmitter) {
                exchange.request().endHandler(new Handler<Void>() {
                    @Override
                    public void handle(Void event) {
                        try {
                            MultiMap res = exchange.request().formAttributes();

                            final String jUsername = res.get("j_username");
                            final String jPassword = res.get("j_password");
```

**Implementation ideas**
We can simply add 4 options to the configuration, to allow the developer specify the following : 
- username/password field names
- post location. 
- locationCookie

### PR DESCRIPTION
**Pull request by mahieddine**
Implement new config items (username/password field name, postLocation and location cookie name) close #14303

### CODE REVIEW

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityRecorder.java
Comment (famod) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityRecorder.java:259]: To me it would make sense to auto-add `/` like for the `*Page` variables.

Comment (mahieddine) [reply extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityRecorder.java]: Yes absolutely, I fixed that in my last commit

**Review by famod — APPROVED** [review]
LGTM, but better leave the final decision up to @stuartwdouglas (or other maintainers of this part of Quarkus).

File: extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityRecorder.java
Comment (sberyozkin) [inline extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpSecurityRecorder.java:260]: @mahieddine Hi, it looks good, minor suggestion - the last 3 local vars are not conditionally initialized - they are just passed further to the mechanism - as such it would be simpler to use `form.usernameFieldName`, etc

**Comment by sberyozkin** [comment]
@mahieddine Thanks for this PR; can you also please add the tests verifying these customizations are effective ?

The other question I have, I'm not really a Form authentication user, but I'm wondering,  when it comes to those 3 properties with the default values starting from `/j_security_check` etc, ``, is it still a standard Form authentication if a post location is starting from something else ? The cookie name seems fine, but not 100% sure about other properties ( - is it problematic that they are hardcoded ?)

Thanks, Sergey

**Review by stuartwdouglas — APPROVED** [review]

**Comment by sberyozkin** [comment]
@mahieddine Can you please add a simple test ? Just copy and paste some existing unit test there :-), and feel free to ignore my other comments (the one regarding the local vars one is subjective).

**Comment by mahieddine** [comment]
@sberyozkin sorry i didn't have much time, i will do it today no problem.  Regarding your question Yes and No, In a pure JSR world [certainly](https://javaee.github.io/servlet-spec/downloads/servlet-4.0/servlet-4_0_FINAL.pdf)

```
In order for the authentication to proceed appropriately, the action of the login form
must always be j_security_check. This restriction is made so that the login form
will work no matter which resource it is for, and to avoid requiring the server to
specify the action field of the outbound form. The login form should specify
autocomplete=”off” on the password form field.
```

But in a "quarkus/vertx" world, correct me if im wrong but i dont think so, otherwise we would also forbid a user to attach a route on '/j_security_check' too ? (btw in another PR i think we should throw a warning or even an error if someone attach a route on the 'postLocation') 

As another argument, In Spring we can customize the form user/pass field names, by setting for e.g 'usernameParameter', 'passwordParameter',...

So all in all i would say giving the user the possibility to customize his form will give him more flexibility to integrate with the existing world without changing the auth flow. Also users who want to stick strictly with the JSR standard in terms of naming are allowed to of course after all its and will remains the default behavior.

**Comment by sberyozkin** [comment]
@mahieddine thanks for the clarification - makes sense. I've restarted the build as some jobs have failed with the unrelated error, will merge once it is green

**Review by sberyozkin — APPROVED** [review]

**Comment by sberyozkin** [comment]
JDK11 Windows build has failed, but the unit test you have added passed:
```
2021-01-20T11:34:47.2785187Z [INFO] Running io.quarkus.vertx.http.security.FormAuthParametersTestCase
2021-01-20T11:34:48.2347673Z 2021-01-20 11:34:48,228 INFO  [io.quarkus] (main) Quarkus 999-SNAPSHOT on JVM started in 0.394s. Listening on: http://localhost:50568
2021-01-20T11:34:48.2349477Z 2021-01-20 11:34:48,228 INFO  [io.quarkus] (main) Profile test activated. 
2021-01-20T11:34:48.2350322Z 2021-01-20 11:34:48,228 INFO  [io.quarkus] (main) Installed features: [cdi, security]
2021-01-20T11:34:49.4728103Z 2021-01-20 11:34:49,457 INFO  [io.quarkus] (main) Quarkus stopped in 0.016s
2021-01-20T11:34:49.4954544Z [INFO] Tests run: 3, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 2.215 s - in io.quarkus.vertx.http.security.FormAuthParametersTestCase
```

**Pull request merged**
